libsupport/include/katana/Cache.h

```diff
@@ -37,7 +43,9 @@ class KATANA_EXPORT Cache {
   /// Construct an LRU cache that has a fixed number of entries.
   Cache(
       size_t capacity,  // number of entries
-      std::function<void(const Key& key)> evict_cb = nullptr)
+      std::function<
+          void(const Key& key, uint64_t approx_bytes, tsuba::RDG* rdg)>
```

Comment:
I'm a little confused about the cardinality relationship between a `Cache` a `PropertyGraph` and an `RDG`. Is it a cache per RDG? or can a cache contain the properties of multiple RDGs? If it's the former (seems like it is) it's a little weird to expect users to keep track of both the cache and which RDG it belongs to.

As a separate point It's a little weird that it's an `RDG` pointer on this general class, why not make it a template variable? Something like:
`template <typename Key, typename Value, typename CallerState>`

If at all possible I would prefer just to capture the `rdg` pointer in a lambda since the callback would be the onlything that cares about it but I suspect you did it this way because there's a construction ordering issue.

Reply:
The cache contains the properties for multiple AtributedGraph instances, which really should be the same RDG, though I do not have a check for this.  Now that I think about it, it has to be the same RDG.  If we want to get into caching property columns across RDGs we need to have checksums or some way of representing in storage that property values are shared.  That is not in scope for this PR and so far as I know is not on our roadmap.

The RDG pointer comes from `tsuba::RDG::DoMake`, which captures the RDG this pointer as it creates a lambda to call `AddProperties`.  I agree with you that having an RDG* parameter in the callback is ugly.  But what else can I do?  The RDG does not exist when I instantiate the cache, so how can I pass a pointer to it at that point?  I used to capture the `this `pointer for QueryClient, but I can get cache misses during RDG construction when the PropertyGraph (and hence AttributedGraph) are not yet created.

Reply:
That's what I meant by
> but I suspect you did it this way because there's a construction ordering issue.

Could we make it c-style caller state with an additional class template argument instead?

Reply:
That was a very interesting idea, thanks!.  I had to define the template parameter as a pointer (i.e., as void* not void) because otherwise importing the RDG definition everywhere a PropertyCache is used gets awkward.

I added the rdg directory as part of the property cache key, thanks for bringing this up.  I also encapsulated the PropertyCacheKey, making it a class instead of a struct.  Such is the way of progress.

Reply:
Nice! I think that all sounds good.